File: Tests/[E2] threadSleepAndWakeup/ThreadSleepAndWakeup.cpp
#include <chrono>
#include <iostream>

#include "mingw.thread.h"

int main() {
    const long long sleepDurationMillis = 10;  // Sleep duration in milliseconds

    auto startTime = std::chrono::high_resolution_clock::now();

    // Create a thread to sleep
    std::thread sleepThread([sleepDurationMillis]() {
        std::this_thread::sleep_for(
            std::chrono::milliseconds(sleepDurationMillis));
    });

    // Wait for the thread to finish sleeping
    sleepThread.join();

    auto endTime = std::chrono::high_resolution_clock::now();

    auto durationMillis = std::chrono::duration_cast<std::chrono::milliseconds>(
                              endTime - startTime)
                              .count();

    std::cout << "C++ Thread Sleep and Wakeup Duration: " << durationMillis
              << " milliseconds" << std::endl;

    return 0;
}

File: Tests/[E2] threadSleepAndWakeup/ThreadSleepAndWakeup.java
public class ThreadSleepAndWakeup {

    public static void main(String[] args) {
        final long sleepDurationMillis = 10; // Sleep duration in milliseconds

        long startTime = System.nanoTime();

        // Create a thread to sleep
        Thread sleepThread = new Thread(() -> {
            try {
                Thread.sleep(sleepDurationMillis);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        });

        sleepThread.start();

        // Wait for the thread to finish sleeping
        try {
            sleepThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        long endTime = System.nanoTime();

        long durationMillis = (endTime - startTime) / 1_000_000; // Convert nanoseconds to milliseconds

        System.out.println("Java Thread Sleep and Wakeup Duration: " + durationMillis + " milliseconds");
    }
}
